chat-android/src/main/java/com/ably/chat/TaskScheduler.kt

```diff
@@ -0,0 +1,69 @@
+package com.ably.chat
+
+import java.util.PriorityQueue
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+
+private class Task(
+    private val priority: Int,
+    val coroutineBlock: suspend CoroutineScope.() -> Any,
+    val result: TaskResult<Any>)
+    : Comparable<Task> {
+        override fun compareTo(other: Task): Int {
+            return this.priority.compareTo(other.priority)
+        }
+
+        suspend fun setResult(res: Result<Any>) {
+            result.channel.send(res)
+        }
+}
+
+class TaskResult<T> {
+    // Size of channel is set to 1. This is to avoid sender getting blocked
+    // because receiver doesn't call receive on the channel
+    internal val channel = Channel<Result<T>>(1)
+
+    suspend fun await(): Result<T> {
+        val result = channel.receive()
+        channel.close()
+        return result
+    }
+}
+
+/**
+ * TaskScheduler schedules given coroutine operation mutually exclusive.
+ * @property scope Uses single threaded dispatcher to avoid thread synchronization issues.
+ */
+class TaskScheduler(private val scope: CoroutineScope) {
```

Comment:
I believe implementation of this class implies single-threaded coroutine context, but there is no restriction in public API, which is not very good. Also it's hard to understand how it will be used later.  I am a bit skeptical about this scheduler which is pretty low-leveled, I would prefer built-in `kotlinx.coroutines` tools. In my head it's something like this  

```kotlin
class RoomLifecycleManager {
  private val internalEvents = Channel<InternalEvent>(Channel.UNLIMITED)
  private val attachmentEvents = Channel<AttachmentEvent>(Channel.UNLIMITED)
  private val releasingEvents = Channel<ReleaseEvent>(Channel.UNLIMITED)

  init { subscribeToIncommingEvents() }  

  fun attach() {
     subscribeToContributorsStateChange()
     dispatch(AttachmentEvent.Attach)
  }
  
  fun detach() {
     dispatch(AttachmentEvent.Detach)
  }

   private fun subscribeToIncommingEvents() {  
     withContext(singleThreadContext) {
         while (active) {
                select {
                    internalEvents.onReceive { event -> 
                        processInternalEvent(event)
                    }
                    attachmentEvents.onReceive { event -> 
                        processAttachmentEvent(event)
                    }
                    releasingEvents.onReceive { event -> 
                        processReleasingEvent(event)
                    }
                }
            }
        }
   }
}
```

Reply:
Official documentation mentions API is experimental -> https://kotlinlang.org/docs/select-expression.html.
Though latest release might have made it stable

Reply:
I will go through the documentation again. This is a good idea to use. 
Also, it fits our requirement of `prioritized` execution according to the spec https://sdk.ably.com/builds/ably/specification/pull/200/chat-features/#:~:text=(-,CHA%2DRL7,-)%20Room%20lifecycle

Reply:
I will create a new PR for this. We will see how complicated this can get. 
The main issue is dealing with `while(active)` loop. 
We need to keep track of room getting released and if there are any pending events as such : (